### PR TITLE
feat(predict_assessments): deploy trained model as Cloud Function

### LINKED ISSUE #12: Create a task to run the current assessment value model
## Motivation

In relatively active property markets such as Philadelphia, the prices of properties in different areas of the city frequently fluctuate. As a result, we will want to re-train our valuation model on a regular basis to ensure that it is capturing the most up to date information about the property market.

> NOTE: In reality, training a model like this would probably not need to happen more frequently than on a quarterly basis. However, for the purposes of this project, you should re-train the model weekly.

## Implementation

The task should either be implemented using a Cloud Function (if the model is implemented in SQL or Python) or using a Cloud Run container (if the model is implemented in R or Python).

The input training data should be prepared in a table named `derived.current_assessments_model_training_data`.

There are a few steps implicit in this task:
1. Download the training data from BigQuery (e.g. using the [`bq` command line tool](https://cloud.google.com/bigquery/docs/reference/bq-cli-reference#bq_extract) or using a client library for [R](https://cran.r-project.org/web/packages/bigrquery/index.html) or [Python](https://pypi.org/project/google-cloud-bigquery/))
2. Train the model using the training data
3. Use the trained model to predict the current assessment value of all (residential) properties in the `core.opa_properties` table (which you will also need to download from BigQuery)
4. Load the predictions back into BigQuery

The resulting prediction values should be loaded into a BigQuery table named `derived.current_assessments`, and should have at least three columns:
* `property_id` -- The OPA number of the property
* `predicted_value` -- The predicted current assessment value of the property
* `predicted_at` -- The date and time that the predicion was run. This will be the same for all properties in the table loaded on the same day/time, but notice that you could update this table every time you run predictions, thus keeping a history of how your predictions across the city change over time.

If you are using R for the model (or, in some cases, Python), you will need to create a Docker container that includes the model code and any dependencies. You can use the [rocker/tidyverse](https://hub.docker.com/r/rocker/tidyverse) image as a base image for R, and install any additional packages you need using `install.packages()`. In this case your _Containerfile_ may look something like:

```docker
FROM gcr.io/google.com/cloudsdktool/cloud-sdk:slim

RUN apt-get update

# Install GDAL for ogr2ogr
RUN apt-get install -y r-base r-base-dev r-cran-tidyverse

COPY install.R .
COPY train-and-predict.R .

# Pre-install the packages during build
RUN Rscript install.R

# Train and predict when starting the container
CMD [ "Rscript", "train-and-predict.R" ]
```

- Refer to issue #6 for more information on running containers in Cloud Run.

**Acceptance criteria:**
- [ ] A Cloud Function or Cloud Run Job that runs the current assessment value model and loads the predictions into a BigQuery table named `derived.current_assessments` with the columns `property_id`, `predicted_value`, and `predicted_at`.
- [ ] The task should be scheduled to run as part of the data pipeline, after the `core.opa_properties` table has been updated with the most recent data.

### PR DESCRIPTION
**Pull request by Mayaano**
# Description
Closes #12 . Builds on #11 (PR #40).

Deploys the trained valuation model as a Cloud Function and wires it into the pipeline. **Predict-only** — uses the pre-trained `eda/model.pkl` as-is, no retraining (team decision, deadline-driven).

Adds:
- `tasks/predict_assessments/` — Cloud Function (gen2, 8GB, 1800s) that loads the model bundle, scores all residential properties, writes `derived.current_assessments`
- predict-assessments deploy block in `tasks/deploy.ps1` (also `Copy-Item eda/model.pkl` step)
- New `predict_assessments` step in `data_pipeline.yaml`, plus moves `current_assessment_bins` + `export_property_tile_info` to a parallel block AFTER predict (they read `derived.current_assessments` — previously racing with `create_training_data`, a latent bug)

## Type of change

- [x] New feature
- [x] Bug fix (workflow ordering)
- [ ] Small fix/change
- [ ] Documentation

## What should the reviewer know?

- `sale_year` set to current year in inference (recency signal for "value now") — one-line swap if you'd rather use last-sale year
- `requirements.txt` pins `scikit-learn==1.8.*` to match Tess's training version
- Bundle contract documented in `tasks/predict_assessments/README.md`
- Out of scope: `generate-assessment-chart-config` deploy block in `deploy.ps1` (~L263) is missing trailing flags — separate fix needed before deploy

## Post-merge follow-ups

- [x] Actions required (specified below)

```pwsh
Copy-Item eda/model.pkl tasks/predict_assessments/model.pkl -Force
gcloud functions deploy predict-assessments --gen2 --runtime=python311 --region=us-east4 --source=tasks/predict_assessments --entry-point=predict_assessments --trigger-http --timeout=1800s --memory=8GB --no-allow-unauthenticated
gcloud workflows deploy data-pipeline --location=us-east4 --source=tasks/workflows/data_pipeline.yaml
gcloud workflows run data-pipeline --location=us-east4
```

### CODE REVIEW

File: tasks/run_current_assessments_model/main.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+from datetime import datetime, timezone
+
+import functions_framework
+import joblib
+import numpy as np
+import pandas as pd
+from google.cloud import bigquery
+
+DIR_NAME = pathlib.Path(__file__).parent
+MODEL_PATH = DIR_NAME / "model.pkl"
```
Comment (tess-vu) [inline tasks/run_current_assessments_model/main.py]: This pickle approach doesn't satisfy Issue #12's retraining requirement. Replace with training logic that pulls from `derived.current_assessments_model_training_data` via BigQuery. See Issue #12, it says *"you should re-train the model weekly."*

File: tasks/run_current_assessments_model/main.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+SQL_PATH = DIR_NAME / "predict_assessments.sql"
+
+# Feature column order MUST match how the model was trained.
+# This is the value of `bundle["model"].feature_names_in_` from the saved bundle.
+FEATURE_COLS = [
```
Comment (tess-vu) [inline tasks/run_current_assessments_model/main.py:41]: Feature list looks good and matches the EDA notebook. Keep this as-is, the trained model should use the same features.

File: tasks/run_current_assessments_model/main.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    encoded = pd.Series(-1, index=df.index, dtype=np.int64)
+    mask_known = df["neighborhood"].isin(known)
+    if mask_known.any():
+        encoded.loc[mask_known] = le.transform(df.loc[mask_known, "neighborhood"])
+    df["neighborhood"] = encoded
+
+    return df
+
+
+def predict_and_write(client, project_id, df, model):
```
Comment (tess-vu) [inline tasks/run_current_assessments_model/main.py]: Output DataFrame needs two additional columns `predicted_value_lower` and `predicted_value_upper` (from quantile models at α=0.1 and α=0.9). Without these `property_tile_info.sql` lines 74–75 will return NULLs and the reviewer UI's MOE bar will be empty. Schema field list also needs updating to include both `FLOAT64` fields.

File: tasks/run_current_assessments_model/main.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        ],
+    )
+    print(f"Writing {len(out):,} predictions to {table}.")
+    job = client.load_table_from_dataframe(out, table, job_config=job_config)
+    job.result()
+    return len(out)
+
+
+@functions_framework.http
+def predict_assessments(request):
```
Comment (tess-vu) [inline tasks/run_current_assessments_model/main.py]: Rename entry point to `run_current_assessments_model` to match `deploy.ps1` line ~302(?) and `data_pipeline.yaml`'s `run_model_dependent_steps` block. Both already reference `run-current-assessments-model`.

File: tasks/run_current_assessments_model/main.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        client = bigquery.Client()
+
+        df = load_features(client, project_id)
+        df = preprocess(df, bundle)
+        n_written = predict_and_write(client, project_id, df, bundle["model"])
+
+        msg = f"Wrote {n_written:,} predictions to derived.current_assessments."
+        print(msg)
+        return (msg, 200)
+
```
Comment (tess-vu) [inline tasks/run_current_assessments_model/main.py:392]: Add a step here after predictions are written to upload feature importances JSON to `gs://musa5090s26-team5-public/configs/model_feature_importances.json`. The reviewer UI fetches this for the "Why this prediction?" panel. Expected shape: `{"model": "...", "trained_at": "...", "features": [{feature, label, importance}, ...]}`.

File: tasks/run_current_assessments_model/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+functions-framework==3.*
+google-cloud-bigquery==3.*
+pandas==2.*
+pyarrow==17.*
+db-dtypes==1.*
+scikit-learn==1.8.*
```
Comment (tess-vu) [inline tasks/run_current_assessments_model/requirements.txt]: Loosen to `scikit-learn==1.*` bc pinning to 1.8 will break when the Cloud Function runtime has a different minor version. Also add `google-cloud-storage==2.*` which is needed for the feature importances upload. `joblib` can be removed since we won't be loading a pickle.

File: tasks/run_current_assessments_model/predict_assessments.sql [hunk 7]
```diff
@@ -0,0 +1,69 @@
+-- Issue #12: Inference query for predict_assessments Cloud Function.
```
Comment (tess-vu) [inline tasks/run_current_assessments_model/predict_assessments.sql]: Nice that this is a separate SQL file, other functions in the codebase follow this pattern (e.g. `current_assessment_bins.sql`, `create_training_data.sql`). Keep this separation. However, the query needs to match the training-time feature engineering. Consider whether imputation should happen in SQL (like the medians CTE approach) or in Python post-query.

File: tasks/predict_assessments/.gitignore
Comment (tess-vu) [inline tasks/predict_assessments/.gitignore:1]: This file can be deleted, with the train at runtime approach there's no `model.pkl` to copy or ignore.

File: tasks/deploy.ps1 [hunk 8]
```diff
@@ -325,6 +325,24 @@ gcloud functions deploy generate-tax-year-chart-config `
     --memory=512MB `
     --no-allow-unauthenticated
 
+# Predict Assessments (Issue #12).
```
Comment (tess-vu) [inline tasks/deploy.ps1]: Drop all `deploy.ps1` changes. The deploy block for this function already exists in `main` under the name `run-current-assessments-model` with the correct flags (`--set-env-vars`, `--memory=8GB`, `--cpu=2`). The `Copy-Item model.pkl` step is also no longer needed.

File: tasks/workflows/data_pipeline.yaml [hunk 9]
```diff
@@ -281,6 +281,7 @@ main:
             septa: "Completed"
             create_training_data: "Completed"
             tax_year_assessment_bins: "Completed"
+            predict_assessments: "Completed"
```
Comment (tess-vu) [inline tasks/workflows/data_pipeline.yaml]: Drop all YAML changes. There's a restructured the workflow with a `run_model_dependent_steps` sequential block that correctly sequences model then bins then chart config then map styling then tile export. Your version is missing `generate-assessment-chart-config` and `generate-map-styling-metadata` steps, which would break those functions.

File: tasks/run_current_assessments_model/README.md
Comment (tess-vu) [inline tasks/run_current_assessments_model/README.md:1]: This README is great, the bundle contract table, feature order documentation, and retraining checklist are neat. Please keep and adapt for the new approach, so update function name to `run-current-assessments-model`, document the 5-column output schema, mention the feature importances upload, and remove pkl-specific instructions.

**Review by tess-vu — CHANGES_REQUESTED** [review]
Thanks for getting Issue #12, the README and model bundle documentation are useful and preprocessing logic is solid, PRs #51–54 were merged into `main` and they introduced downstream dependencies that this PR needs to adjust to. Here's what needs to change and why.

### Rename: `predict_assessments` to `run_current_assessments_model`

deploy.ps1 (line ~300) and data_pipeline.yaml (the `run_model_dependent_steps` block) both reference `run-current-assessments-model` as the function name and `tasks/run_current_assessments_model` as the source directory. The folder, entry point, and all references need to use this name, otherwise deploy and the workflow will do an error 404.

### Train at runtime, don't load a pkl

Issue #12 explicitly requires weekly retraining:
> "We will want to re-train our valuation model on a regular basis… you should re-train the model weekly."

This approach loads a static `model.pkl` and only predicts so this means the model never updates when new sales data arrives. The function needs to:
1. Pull training data from `derived.current_assessments_model_training_data` (already built by the `create-training-data` function upstream in the pipeline).
2. Train a GradientBoostingRegressor.
3. Predict on all residential properties.
4. Write results to `derived.current_assessments`.

This also means the `model.pkl` copy step in deploy.ps1, the .gitignore, and the `joblib` dependency can all be removed.

### Add `predicted_value_lower` and `predicted_value_upper` columns.

The merged tile export SQL (property_tile_info.sql, lines 74–75) reads:
```sql
ca.predicted_value_lower AS current_assessed_value_lower,
ca.predicted_value_upper AS current_assessed_value_upper,
```

And the reviewer UI popup (`ui/reviewer/index.html`, the `populatePropertyPopup` function) uses `current_assessed_value_lower` / `current_assessed_value_upper` to show an 80% confidence range bar that could be helpful to flag tax assessors reviewing predictions. If these columns don't exist in `derived.current_assessments`, both the tile export and the popup MOE bar will break.

To produce these, train two additional quantile GBR models (`loss="quantile"`, α=0.1 and α=0.9) alongside the main model, and clamp so `lower ≤ predicted ≤ upper`.

The output table schema should be:
| Column | Type |
|---|---|
| `property_id` | `STRING` |
| `predicted_value` | `FLOAT64` |
| `predicted_value_lower` | `FLOAT64` |
| `predicted_value_upper` | `FLOAT64` |
| `predicted_at` | `TIMESTAMP` |

### Add feature importances JSON upload.

Reviewer UI (`ui/reviewer/index.html`, line ~343) fetches:
```
https://storage.googleapis.com/musa5090s26-team5-public/configs/model_feature_importances.json
```
This goes into the "Why this prediction?" explainability panel in property popups. After training, extract `model.feature_importances_`, pair with human-readable labels, sort descending, and upload to `gs://musa5090s26-team5-public/configs/model_feature_importances.json`. Expected shape should be:
```json
{
  "model": "GradientBoostingRegressor",
  "trained_at": "2026-04-26T...",
  "features": [
    {"feature": "total_livable_area", "label": "Living area (sqft)", "importance": 0.35},
    ...
  ]
}
```
Without this file the UI shows "Feature importances not loaded yet") but the explainability panel is something that was briefly discussed with Mjumbe.

### Don't modify workflow YAML

`data_pipeline.yaml` was restructured in PR #54. The `run_model_dependent_steps` block already sequences the model to bins to chart config to map styling to tile export correctly. I will drop the YAML changes from this PR, just make sure the function name matches what the YAML calls (`run-current-assessments-model`).

### Don't modify `deploy.ps1`

The deploy block for `run-current-assessments-model` was added in PR #54 (with `--set-env-vars`, `--memory=8GB`, `--cpu=2`). I will drop all `deploy.ps1` changes from this PR, including the `Copy-Item model.pkl` step which is not needed as #12 states the model needs to be retrained weekly.

### Keep the README

The `README.md` documenting the model bundle contract, feature order, and retraining checklist is great. Adapt it for the new approach that trains at runtime instead of static pkl, updated output schema, new file name.

---

**Files to Change:**
| Done? | File | Action |
|---|---|---|
| [x] by Tess | `tasks/predict_assessments/` | Rename folder to `tasks/run_current_assessments_model/` |
| [ ] | `main.py` | Replace pickle-load with train-at-runtime logic, so `load_training_data()`, `prepare_training_frame()`, `train_models()` (3 models), `encode_prediction_frame()`, `predict_and_write()` (5 columns), `upload_feature_importances()` |
| [ ] I can do this one since `main.py` changes are substantial and I have full pipeline context? Let me know | `predict_assessments.sql` | May keep as separate SQL file but update to match new approach |
| [x] by Tess | `requirements.txt` | Add `google-cloud-storage`, remove `joblib`, unpin sklearn |
| [ ] | `README.md` | Update for new function name, output schema, no pkl |
| [x] by Tess | `.gitignore` | Delete (no pkl to ignore) |
| [x] by Tess | `deploy.ps1` changes | Drop entirely, already handled in main branch |
| [x] by Tess | `data_pipeline.yaml` changes | Drop entirely, already handled in main branch |

**Comment by tess-vu** [comment]
main.py and README.md:
- Updated FEATURE_LABELS to match features in `create_training_data.sql`, I made a mistake and thought that the features were fine, but I forgot to check the training table, sorry about that!

predict_assessments.sql:
- Dead code and deleted.

**Review by tess-vu — APPROVED** [review]
